### PR TITLE
MAINT: utility module for Intel data parallel libs; import checks in one place

### PR DESCRIPTION
**Pull request by samir-nasibli**
### Description
These changes centralized data parallel libs checks, such as DPCtl and DPNP availability. (Also discussed on https://github.com/intel/scikit-learn-intelex/pull/1861#discussion_r1738343933)
Proposed changes:
* doing the check of the data parallel libs availability in one place, reusing related variable through the lib.
* added `is_dpctl_available`/`is_dpnp_availble` tools for checking the specific dpctl, dpnp versions
* renamed `is_dpctl_available` from `onedal.tests.utils._device_selection` module to `is_dpctl_device_available`. This tool is used in `sklearnex`/`onedal` tests where it is needed to check required device(s) availability via `dpctl`.
* Comments and docstring is added.

Non prior. Should be integrated after main is unfreeze.

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes, if necessary (updated in # - _add PR number_)
- [x] The unit tests pass successfully.
- [x] I have run it locally and tested the changes extensively.
- [x] I have resolved any merge conflicts that might occur with the base branch.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_
- [x] I have added a respective label(s) to PR if I have a permission for that.

### CODE REVIEW

File: onedal/tests/utils/_device_selection.py [hunk 1]
```diff
@@ -62,9 +64,7 @@ def get_memory_usm():
 
 
 def is_dpctl_available(targets=None):
-    try:
-        import dpctl
-
+    if dpctl_available:
```
Comment (samir-nasibli) [inline onedal/tests/utils/_device_selection.py:67]: the primitive on line 66 probably should be renamed.

Comment (samir-nasibli) [reply onedal/tests/utils/_device_selection.py:67]: Done

File: onedal/utils/_dppy_available.py
Comment (samir-nasibli) [inline onedal/utils/_dppy_available.py:1]: Renaming of the file is need.

File: onedal/tests/utils/_device_selection.py [hunk 2]
```diff
@@ -61,20 +63,17 @@ def get_memory_usm():
         return []
 
 
-def is_dpctl_available(targets=None):
-    try:
+def is_dpctl_device_available(targets):
+    if dpctl_available:
```
Comment (icfaust) [inline onedal/tests/utils/_device_selection.py:67]: if dpctl_available check can be made an import-time check, rather than runtime.

File: onedal/utils/_dppy_available.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    """Checks availability of DPCtl package"""
+    try:
+        import dpctl
+        import dpctl.tensor as dpt
+
+        dpctl_available = True
+    except ImportError:
+        dpctl_available = False
+    if dpctl_available and not version == "":
+        dpctl_available = dpctl.__version__ >= version
```
Comment (icfaust) [inline onedal/utils/_dppy_available.py:30]: This will fail/possibly operate funny when characters exist in the version (see "0.19.0dev0" https://intelpython.github.io/dpctl/latest/index.html) especially if a bleeding edge dpctl version is used. Try to follow sklearn_check_version for proper version checks.

File: onedal/utils/_dppy_available.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+def is_dpnp_available(version=""):
+    """Checks availability of DPNP package"""
+    try:
+        import dpnp
+
+        dpnp_available = True
+    except ImportError:
+        dpnp_available = False
+    if dpnp_available and not version == "":
+        dpnp_available = dpnp.__version__ >= version
```
Comment (icfaust) [inline onedal/utils/_dppy_available.py:43]: Similar change.

**Review by icfaust — COMMENTED** [review]
I'd like a change of the name of "dppy", which I can only see in reference to dpctl's official/unofficial anaconda channel.  Something which references data parallel intel or SYCL would be preferred.

File: onedal/utils/_dp_helpers.py
Comment (samir-nasibli) [inline onedal/utils/_dp_helpers.py:1]: https://intelpython.github.io/DPEP/main/index.html

```
onedal/utils/_dpep_helpers.py
```

File: onedal/utils/_dpep_helpers.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""Utilities for Data Parallel Extensions libs, such as DPNP, DPCtl"""
+
+from daal4py.sklearn._utils import _package_check_version
+
+
+def is_dpctl_available(version=None):
```
Comment (Alexsandruss) [inline onedal/utils/_dpep_helpers.py:22]: These functions need caching just in case they will be used more extensively in future.

File: onedal/utils/_dpep_helpers.py
Comment (Alexsandruss) [inline onedal/utils/_dpep_helpers.py:1]: What does `dpep` mean? I believe it is confusing name for helper file.

Comment (samir-nasibli) [reply onedal/utils/_dpep_helpers.py:1]: https://github.com/IntelPython/DPEP
https://intelpython.github.io/DPEP/main/
Do you want me provide links in the docs of the module or just rename?

Comment (samir-nasibli) [reply onedal/utils/_dpep_helpers.py:1]: I have provided small description in the top of the module

Comment (Alexsandruss) [reply onedal/utils/_dpep_helpers.py:1]: `data_parallel_extensions_helper` would be better.

File: onedal/tests/utils/_device_selection.py (hunk 2 above)
Comment (samir-nasibli) [reply onedal/tests/utils/_device_selection.py:67]: moving out this boolean check doesn't give a lot, but moving out package import make sense. Going to move them out from primitives.

File: onedal/utils/_dpep_helpers.py
Comment (samir-nasibli) [reply onedal/utils/_dpep_helpers.py:1]: I think shorter name for the module is ok. I added links in module docstring, that should make this clear. https://github.com/intel/scikit-learn-intelex/pull/1936/commits/0d44141034fab3d306b6547bf5519bb003bff4f6

**Review by Alexsandruss — APPROVED** [review]

**Review by samir-nasibli — COMMENTED** [review]
Added ticket for use of this module in the future dev guidelines

**Review by icfaust — COMMENTED** [review]
Great PR, I think its go to go, and is well written.  Stylistically, I would swap "_dpep_helpers" for something like "_usm_array_api", "_usm_helpers", "_usm_utils", or something just for consistency/ if we add additional usm features later on that we want to centralize.
